### ISSUE TITLE
[Feature] Allow custom publish subdirectory

### ISSUE THREAD

**Issue by jacobwgillespie** [lerna/lerna#91]
**EDIT:** I overlooked `package/[name]/scripts/prepublish.js`, so that's awesome.  It would however be a bit difficult to customize the directory in which `npm publish` is run just with a prepublish script.  I'll adjust this issue to recommend adding a way to customize that publish directory, since it's hard-coded in `publishTaggedInDir`.

---

It would be awesome to be able to customize the publish behavior (like the publish directory) - for my use case, I build the packages like `packages/abc/src` into a directory called `packages/abc/dist`, then copy the package.json into `dist` and publish from there (to have a flat module structure).  I've gotten it working for now by overriding the exports of `npmUtils`, but in the future it would be awesome to be able to customize this step:

``` javascript
#!/usr/bin/env node

// Wrapper script that patches lerna with our custom publish path

import child from 'child_process';
import logger from 'lerna/lib/utils/logger';
import path from 'path';

const newPublishTaggedInDir = logger.logifyAsync('npmUtils.publishTaggedInDir', (tag, dir, callback) => {
  const cmd = `cd ${dir}/dist && cp ../package.json . && npm publish --tag ${tag}`;
  child.exec(cmd, (err, stdout, stderr) => {
    callback(stderr || err, stdout);
  });
});

// Manually require npmUtils to load it into the cache
require('lerna/lib/utils/npmUtils');

// Patch the exports
const modulePath = path.resolve(__dirname, '../node_modules/lerna/lib/utils/npmUtils.js');
require('module')._cache[modulePath].exports.publishTaggedInDir = newPublishTaggedInDir;

// Run the original binary
require('lerna/bin/lerna');
```

**Comment by thejameskyle** [comment]
You can put package-specific scripts inside `packages/:package_name/scripts/prepublish.js` or `packages/:package_name/scripts/postpublish.js`. If you want to do something before lerna entirely, I recommend just running it before you run lerna with like an npm script or whatever.

**Comment by jacobwgillespie** [comment]
Ah, nice, I overlooked the pre/post publish scripts.  I'm already doing stuff outside of lerna before running publish (with a Makefile).

I edited the title and issue description now to be specific about customizing the publish directory, as it would still be pretty difficult to change the directory from a prepublish script alone (it's kinda hard-coded in `publishTaggedInDir`).

**Comment by thejameskyle** [comment]
Sorry I left this for so long, I'm not sure what you are asking for with customizing the publish directory. Do you mean running `npm publish` in a nested directory within each package?

**Comment by nadavsinai** [comment]
@thejameskyle  this would be very beneficial for us too - as you said - to create flat packages when publishing

**Comment by jacobwgillespie** [comment]
> Do you mean running npm publish in a nested directory within each package?

Sorry, super late reply, but yes.  The idea is that the `npm publish` would run inside a subdirectory, so like `packages/package-name/dist` as that would give more control over what appeared in the final published package beyond a basic `.npmignore`, for example a customized `package.json` or README different from the parent.

**Comment by nadavsinai** [comment]
@jacobwgillespie  thanks for your reply, I actually used (and updated) your hacky function replacement code to try to make it work, unfortunately I got stuck,I ended up creating a static index.js which states this funny thing :
````
 export * from './dist';
```
and this way the imports from my lib don't need to drill further down then the 'root', setting the main field only didn't work.
Now I can get away with that as I am exporting an es6 module syntax (this is internal NPM repo, all code is eventually consumed by webpack2) not many people are in that position (yet), in Es5/CJS it is not so elegant and publishing in a sub folder is much nicer.

**Comment by jacobwgillespie** [comment]
@nadavsinai in case you ever need to do that with standard CommonJS, it is possible with:

```javascript
module.exports = require('./dist');
```

ES6 isn't necessarily required there.

**Comment by evocateur** [comment]
@jacobwgillespie What you describe is quite possible in stock npm (which is what lerna strives to emulate when it comes to publishing): a `prepublish` script to generate the artifacts and move them in-place, and a `postpublish` script to clean up the generated artifacts. I don't think you need to run `npm publish` in a different directory.

Or, even simpler, just configure `"main": "dist/index.js"` in `package.json`? Use the [files](https://docs.npmjs.com/files/package.json#files) array to customize what is included in the tarball, perhaps.

**Comment by jacobwgillespie** [comment]
@evocateur you are correct, however in my case we wanted an entirely new package.json inside the resultant build.  From what I can tell, that is the only file you cannot change through some combination of options during publishing without resorting to an entirely new directory.

**Comment by evocateur** [comment]
How does this new package.json differ from the original?

**Comment by jacobwgillespie** [comment]
At this point I'm no longer working on the project and can't remember exactly what we were trying to do. 🙂  It was something like having a custom set of dependencies and devDependencies or a different value for engines or something similar.

**Comment by evocateur** [comment]
In general, we try to stay consistent with common `npm` idioms, such as honoring the `package.json` as the source of truth for "how to express this module's dependencies". We're not likely to implement this extension.

**Comment by rkusuma** [comment]
@jacobwgillespie where is the `package/[name]/scripts/prepublish.js`?
i'm encounter the problem like you.. i need to publish the `dist` folder instead of current folder in each `packages`

**Comment by sebelga** [comment]
Hi,
I am also interested in being able to run the "npm publish" from a sub directory. I think it is a common scenario to have /src folder and a building script creating a /dist folder with its package.json and bundles. 

What would be the current solution to only publish the "dist/" folder generated?

thanks!

**Comment by jacobwgillespie** [comment]
@rkusuma that was referring to https://github.com/lerna/lerna/issues/91#issuecomment-201488723 - I don't believe that lets you publish inside a subfolder though.

@sebelga if you're okay with the `package.json` not being directly inside the `dist` folder, you can omit all the folders / files except for `dist` with the `files` setting inside `package.json`:

```json
{
  "name": "my-package",
  "main": "dist/index.js",
  "files": [
    "dist",
    "LICENSE"
  ]
}
```

That way your published package will be `package.json`, `README.md`, `LICENSE`, and the `dist` folder.

**Comment by evocateur** [comment]
A `files` array with a generated `main` is the way to go. If you need a "flat" ES module, maybe instead use a rollup bundle in the `module` field and tree-shake?

**Comment by sebelga** [comment]
@jacobwgillespie thanks I will give that a try and let you know.

**Comment by Tom910** [comment]
Some problem. I want to receive 'flat' structure modules

**Comment by sebelga** [comment]
@jacobwgillespie thanks for the "files" tip, it works perfectly 👍

**Comment by bennyn** [comment]
At [Wire](https://wire.com/) we are also struggling with releasing from a subdirectory with Lerna. 

Before Lerna we were doing `cd ./dist && npm publish` when `npm publish` got invoked. We also executed a prepublish script which adjusted our package.json file and copied it to our `dist` folder, so that references like the main entry point in `dist/package.json` did not became invalid:

```js
const fs = require('fs');

const originalPackage = require('../package.json');
originalPackage.main = './index.js';
delete originalPackage.scripts;

fs.writeFileSync('./dist/package.json', JSON.stringify(originalPackage, null, '  '));
```

Unfortunately this `cd ./dist && npm publish` trick doesn't work anymore with `lerna publish`. That's why we had to rewrite our component exports to make it fit with flat structures: https://github.com/wireapp/wire-web-packages/pull/196
